libs/payments/cart/src/lib/manager.ts

```diff
@@ -0,0 +1,113 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+import { v4 as uuidv4 } from 'uuid';
+import { Cart } from '../../../../shared/db/mysql/account/src';
+import { uuidTransformer } from '../../../../shared/db/mysql/core/src';
+import { Logger } from '../../../../shared/log/src';
+import { UpdateCartInput } from '../gql/update-cart.input';
+import { CartState, SetupCart } from './types';
+import { Cart as CartType } from '../gql';
+import { InvoiceFactory } from './factories';
+
+const DEFAULT_INTERVAL = 'monthly';
+
+export class CartManager {
+  private log: Logger;
+  constructor(log: Logger) {
+    this.log = log;
+  }
+
+  public async setupCart(input: SetupCart): Promise<CartType | null> {
+    const id = uuidv4({}, Buffer.alloc(16)).toString('hex');
+    const currentDate = Date.now();
+    const cart = await Cart.query().insert({
+      ...input,
+      id,
+      state: CartState.START,
+      interval: input.interval || DEFAULT_INTERVAL,
+      createdAt: currentDate,
+      updatedAt: currentDate,
+    });
+
+    return {
+      ...cart,
+      nextInvoice: InvoiceFactory(), // Temporary
+    };
+  }
+
+  public async restartCart(cartId: string): Promise<CartType | null> {
+    const id = uuidTransformer.to(cartId);
+    await Cart.query()
+      .update({ state: CartState.START, updatedAt: Date.now() })
+      .where('id', id);
+
+    const cart = await Cart.query().where('id', id).first();
+
+    if (!cart) {
+      return null;
+    }
+
+    return {
+      ...cart,
+      nextInvoice: InvoiceFactory(), // Temporary
+    };
+  }
+
+  public async checkoutCart(cartId: string): Promise<CartType | null> {
+    // QUESTION
+
+    // I could not get this code to work. Any idea what I might be doing wrong?
+    // Followed this example here.
+    // https://vincit.github.io/objection.js/api/query-builder/mutate-methods.html#updateandfetch
+
+    // const cart = await Cart.query()
+    //   .where('id', uuidTransformer.to(cartId))
+    //   .first();
+    // if (!cart) {
+    //   return null;
+    // }
+    // const updateResult = await cart
+    //   .$query()
+    //   .updateAndFetch({ state: CartState.PROCESSING });
```

Comment:
For the life of me I couldn't get `updateAndFetch` working yesterday. `updateAndFetchById` also doesn't seem to work, since we need to `uuidTransform.to(` on the ID.

Any insight or ideas would be appreciated.

Reply:
Updated this to `patchAndFetch`. The docs recommend `patch` over `upgrade`. I left a comment and link to the docs.